core/emulate.c

```diff
@@ -1105,6 +1105,8 @@ em_status_t EMCALL em_emulate_insn(struct em_context_t *ctxt)
         if (READ_GPR(REG_RCX, ctxt->address_size) == 0) {
             goto done;
         }
+        // Update memory operands
+        decode_operands(ctxt);
```

Comment:
One way to fix the black screen issue is to move this line back to where it was (before every `goto restart`), but that's suboptimal. There must be something in `decode_operands` that makes it unsafe to be called at the first iteration of REP...

Reply:
I think the problem is that `decode_operands` clears `em_operand_t::flags` for destination and source operands, which can be nonzero (e.g. `OP_READ_PENDING`) when we re-enter `em_emulate_insn` for the same iteration of the same REP instruction, due to a previous `operand_read/write` attempt that incurred an exit to QEMU. In the case of re-entrance, there's no need to update `operand_mem_t::ea`, since DI/SI has not changed.

Reply:
But it's okay to clear those flags toward the end of `em_emulate_insn`, because at that point we have completed emulating the instruction (or one iteration of it).

Reply:
In fact, we must make sure those operand flags are reset before jumping to `restart`, or `operand_read/write` will end up doing nothing. So we still want to call `decode_operands()` right before `goto restart`.

Sorry for the fuss! I've made a patch to fix it and will push it shortly. I hope it still offers a slight improvement over the original logic.